### PR TITLE
Frecency implementation

### PR DESCRIPTION
**Pull request by JoshuaWilkes**

### CODE REVIEW

**Comment by JoshuaWilkes** [comment]
USe the debug profile for frecency to run the demo, and you can play with teh weights by changing teh value in code where marked

File: cmd/assume/main.go [hunk 1]
```diff
@@ -171,21 +172,35 @@ func AssumeCommand(c *cli.Context) error {
...
+		}
+
+		// Run this save method async so that is doesn't block the cli
+		go func() { _ = frecentReasons.Save(*reason) }()
 	} else if !matchedRole.Audited && reason != nil {
 		fmt.Fprintf(os.Stderr, "This role is not audited (https://granted.dev/audited-roles). The reason you've provided will be ignored.\n")
 	}
 
+	// @TODO save the latest reason here
+
```
Comment (jordiup) [inline cmd/assume/main.go:203]: Flagging this one in case you meant to add another method here?

Comment (jordiup) [reply cmd/assume/main.go:203]: Seems like it's not blocking though

File: pkg/frecency/frecency.go [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+		r.FrequencyScore = math.Log10(float64(r.Frequency) / float64(store.MaxFrequency) * 10)
+		lastUsedDiff := float64(r.LastUsed.Sub(store.OldestDate))
+		if lastUsedDiff > 0 {
+			r.LastUsedScore = math.Log10(lastUsedDiff / float64(time.Since(store.OldestDate)) * 10)
+		} else {
+			r.LastUsedScore = 1
+		}
+		r.FrecencySortingScore = r.FrequencyScore*FrequencyWeight + r.LastUsedScore*DateWeight
+	}
+
```
Comment (jordiup) [inline pkg/frecency/frecency.go:135]: 🔥

**Review by jordiup — APPROVED** [review]
Nice work Josh, the algo is awesome 

Feel free to merge when you're ready 👌

File: cmd/assume/main.go [hunk 3]
```diff
@@ -171,21 +172,35 @@ func AssumeCommand(c *cli.Context) error {
...
+			reason = &reasonResponse
+		}
+
+		// Run this save method async so that is doesn't block the cli
+		go func() { _ = frecentReasons.Save(*reason) }()
 	} else if !matchedRole.Audited && reason != nil {
 		fmt.Fprintf(os.Stderr, "This role is not audited (https://granted.dev/audited-roles). The reason you've provided will be ignored.\n")
 	}
 
+	// @TODO save the latest reason here
```
Comment (JoshuaWilkes) [inline cmd/assume/main.go:202]:
```suggestion
```

**Pull request merged**